File: app/src/main/java/za/co/woolworths/financial/services/android/util/QueryBadgeCounter.java
package za.co.woolworths.financial.services.android.util;

import java.util.List;
import java.util.Observable;

import retrofit2.Call;
import za.co.woolworths.financial.services.android.contracts.IResponseListener;
import za.co.woolworths.financial.services.android.models.dto.CartSummary;
import za.co.woolworths.financial.services.android.models.dto.CartSummaryResponse;
import za.co.woolworths.financial.services.android.models.dto.MessageResponse;
import za.co.woolworths.financial.services.android.models.dto.VoucherCollection;
import za.co.woolworths.financial.services.android.models.dto.VoucherResponse;
import za.co.woolworths.financial.services.android.models.network.CompletionHandler;
import za.co.woolworths.financial.services.android.models.network.OneAppService;

import static za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity.INDEX_ACCOUNT;
import static za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity.INDEX_CART;
import static za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity.INDEX_REWARD;

public class QueryBadgeCounter extends Observable {

    private int cartCount;
    private int voucherCount;
    private int messageCount;
    private int updateAtPosition;

    private Call<MessageResponse> mGetMessage;
    private Call<VoucherResponse> mGetVoucher;
    private Call<CartSummaryResponse> mGetCartCount;

    private static QueryBadgeCounter INSTANCE = null;

    // Returns a single instance of this class, creating it if necessary.
    public static QueryBadgeCounter getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new QueryBadgeCounter();
        }
        return INSTANCE;
    }

    public void setCartCount(int count, int updateAtPosition) {
        this.cartCount = count;
        this.updateAtPosition = updateAtPosition;
        setChanged();
        notifyObservers();
    }

    private void setVoucherCount(int count, int updateAtPosition) {
        this.voucherCount = count;
        this.updateAtPosition = updateAtPosition;
        setChanged();
        notifyObservers();
    }

    private void setMessageCount(int count, int updateAtPosition) {
        this.messageCount = count;
        this.updateAtPosition = updateAtPosition;
        setChanged();
        notifyObservers();
    }

    public void notifyBadgeCounterUpdate(int updateAtPosition) {
        this.updateAtPosition = updateAtPosition;
        setChanged();
        notifyObservers();
    }

    public int getCartCount() {
        return cartCount;
    }

    public int getVoucherCount() {
        return voucherCount;
    }

    public int getMessageCount() {
        return messageCount;
    }

    public int getUpdateAtPosition() {
        return updateAtPosition;
    }

    public void queryAllBadgeCounters() {
        queryVoucherCount();
        queryCartSummaryCount();
        queryMessageCount();
    }

    public void queryMessageCount() {
        if (isUserAuthenticated()) return;
        if (isC2User()) return;
        mGetMessage = loadMessageCount();
    }

    private boolean isC2User() {
        return !SessionUtilities.getInstance().isC2User();
    }

    public void queryVoucherCount() {
        if (isUserAuthenticated()) return;
        if (isC2User()) return;
        mGetVoucher = loadVoucherCount();
    }

    private boolean isUserAuthenticated() {
        return !SessionUtilities.getInstance().isUserAuthenticated();
    }

    public void queryCartSummaryCount() {
        if (isUserAuthenticated()) return;
        mGetCartCount = loadShoppingCartCount();
    }

    private Call<VoucherResponse> loadVoucherCount() {
        Call<VoucherResponse> voucherCall = OneAppService.INSTANCE.getVouchers();
        voucherCall.enqueue(new CompletionHandler<>(new IResponseListener<VoucherResponse>() {
            @Override
            public void onSuccess(VoucherResponse voucherResponse) {
                switch (voucherResponse.httpCode) {
                    case 200:
                        VoucherCollection voucherCollection = voucherResponse.voucherCollection;
                        if (voucherCollection != null) {
                            if (voucherCollection.vouchers != null) {
                                setVoucherCount(voucherCollection.vouchers.size(), INDEX_REWARD);
                            }
                        }
                        break;
                    default:
                        break;
                }
            }

            @Override
            public void onFailure(Throwable error) {

            }
        }, VoucherResponse.class));

        return voucherCall;
    }

    private Call<CartSummaryResponse> loadShoppingCartCount() {
        GetCartSummary cartSummary = new GetCartSummary();
        return cartSummary.getCartSummary(new IResponseListener<CartSummaryResponse>() {
            @Override
            public void onSuccess(CartSummaryResponse cartSummaryResponse) {
                if (cartSummaryResponse.httpCode == 200) {
                    if (cartSummaryResponse.data == null) return;
                    List<CartSummary> cartSummary = cartSummaryResponse.data;
                    if (cartSummary.get(0) != null) {
                        setCartCount(cartSummary.get(0).totalItemsCount, INDEX_CART);
                    }
                }
            }

            @Override
            public void onFailure(Throwable error) {

            }
        });
    }

    private Call<MessageResponse> loadMessageCount() {
        Call<MessageResponse> messageResponseCall = OneAppService.INSTANCE.getMessagesResponse(5, 1);
        messageResponseCall.enqueue(new CompletionHandler<>(new IResponseListener<MessageResponse>() {
            @Override
            public void onSuccess(MessageResponse messageResponse) {
                setMessageCount(messageResponse.unreadCount, INDEX_ACCOUNT);
            }

            @Override
            public void onFailure(Throwable error) {

            }
        }, MessageResponse.class));
        return messageResponseCall;
    }

    public void clearBadge() {
        setCartCount(0, INDEX_CART);
        setMessageCount(0, INDEX_ACCOUNT);
        setVoucherCount(0, INDEX_REWARD);
    }

    public void cancelCounterRequest() {
        cancelRequest(mGetMessage);
        cancelRequest(mGetVoucher);
        cancelRequest(mGetCartCount);
    }

    private void cancelRequest(Call call) {
        if (call != null && !call.isCanceled()) {
            call.cancel();
        }
    }

    @Override
    public boolean hasChanged() {
        return true; //super.hasChanged();
    }

    public void queryBadgeCount() {
        updateAtPosition = 10;
        setChanged();
        notifyObservers();
    }
}

File: app/src/main/java/za/co/woolworths/financial/services/android/util/QueryBadgeCounter.kt
package za.co.woolworths.financial.services.android.util

import retrofit2.Call
import za.co.woolworths.financial.services.android.contracts.IGenericAPILoaderView
import za.co.woolworths.financial.services.android.contracts.IResponseListener
import za.co.woolworths.financial.services.android.models.dto.CartSummary
import za.co.woolworths.financial.services.android.models.dto.CartSummaryResponse
import za.co.woolworths.financial.services.android.models.dto.MessageResponse
import za.co.woolworths.financial.services.android.models.dto.VoucherCount
import za.co.woolworths.financial.services.android.models.network.OneAppService
import za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity
import za.co.woolworths.financial.services.android.ui.extension.cancelRetrofitRequest
import za.co.woolworths.financial.services.android.ui.extension.request
import java.util.Observable

class QueryBadgeCounter : Observable() {
    var cartCount = 0
        private set
    var voucherCount = 0
        private set
    var messageCount = 0
        private set
    var updateAtPosition = 0
        private set

    private var isCartSummaryFailed = true

    private var mGetMessage: Call<MessageResponse>? = null
    private var mGetVoucher: Call<VoucherCount>? = null
    private var mGetCartCount: Call<CartSummaryResponse>? = null

    fun setCartCount(count: Int) {
        cartCount = count
        this.updateAtPosition = BottomNavigationActivity.INDEX_CART
        notifyUpdate()
    }
    fun getCartItemCount(): Int {
        return cartCount
    }

    private fun setVoucherCount(count: Int) {
        voucherCount = count
        this.updateAtPosition = BottomNavigationActivity.INDEX_REWARD
        notifyUpdate()
    }

    private fun setMessageCount(count: Int) {
        messageCount = count
        this.updateAtPosition = BottomNavigationActivity.INDEX_ACCOUNT
        notifyUpdate()
    }

    fun notifyBadgeCounterUpdate(updateAtPosition: Int) {
        this.updateAtPosition = updateAtPosition
        notifyUpdate()
    }

    fun queryMessageCount() {
        if (!isUserAuthenticated) return
        if (!isC2User) return
        mGetMessage = loadMessageCount()
    }

    private val isC2User: Boolean
        get() = SessionUtilities.getInstance().isC2User

    fun queryVoucherCount() {
        if (!isUserAuthenticated) return
        if (!isC2User) return
        mGetVoucher = loadVoucherCount()
    }

    private val isUserAuthenticated: Boolean
        get() = SessionUtilities.getInstance().isUserAuthenticated

    fun queryCartSummaryCount() {
        if (!isUserAuthenticated) return
        mGetCartCount = loadShoppingCartCount()
    }

    fun updateCartSummaryCount() {
        if (!isUserAuthenticated) return
        if(isCartSummaryFailed){
            queryCartSummaryCount()
            return
        }
        this.updateAtPosition = BottomNavigationActivity.INDEX_CART
        notifyUpdate()
    }

    private fun loadVoucherCount(): Call<VoucherCount>? {
        return request(OneAppService().getVouchersCount(), object : IGenericAPILoaderView<Any> {
            override fun onSuccess(response: Any?) {
                (response as? VoucherCount)?.apply {
                        when (httpCode) {
                            200 -> count?.let { count -> setVoucherCount(count) }
                            else -> setVoucherCount(0)
                        }
                }
            }
        })
    }

    private fun loadShoppingCartCount(): Call<CartSummaryResponse>? {

        return GetCartSummary().getCartSummary(object : IResponseListener<CartSummaryResponse> {
            override fun onSuccess(response: CartSummaryResponse?) {
                when (response?.httpCode) {
                    200 -> {
                        response.data.getOrNull(0)?.apply {
                            if (totalItemsCount != null)
                                setCartCount(totalItemsCount)
                        }
                    }
                }
            }
        })
    }

    private fun loadMessageCount(): Call<MessageResponse>? {
        return request(OneAppService().getMessagesResponse(5, 1), object : IGenericAPILoaderView<Any> {
            override fun onSuccess(response: Any?) {
                (response as? MessageResponse)?.unreadCount?.let { unreadCount ->
                    setMessageCount(unreadCount)
                }
            }
        })
    }

    fun clearBadge() {
        setCartCount(0)
        setMessageCount(0)
        setVoucherCount(0)
    }

    fun cancelCounterRequest() {
        cancelRetrofitRequest(mGetMessage)
        cancelRetrofitRequest(mGetVoucher)
        cancelRetrofitRequest(mGetCartCount)
    }

    override fun hasChanged(): Boolean {
        return true
    }

    fun queryBadgeCount() {
        updateAtPosition = 10
        notifyUpdate()
    }

    private fun notifyUpdate() {
        setChanged()
        notifyObservers()
    }

    fun setCartSummaryResponse(cartSummary: CartSummary?, isFailedCartSummary: Boolean = false) {
        cartCount = cartSummary?.totalItemsCount ?: 0
        isCartSummaryFailed = isFailedCartSummary
    }

    companion object {
        private var INSTANCE: QueryBadgeCounter? = null

        // Returns a single instance of this class, creating it if necessary.
        @JvmStatic
        val instance: QueryBadgeCounter
            get() {
                if (INSTANCE == null) {
                    INSTANCE = QueryBadgeCounter()
                }
                return INSTANCE as QueryBadgeCounter
            }
    }
}